research/json-crdt.md

```diff
@@ -0,0 +1,134 @@
+
+### Operation Representation
+
+An operation is a tuple of the form
+
+```
+op (
+	id: N x ReplicaID
+	deps: P(N x ReplicaID)
+	cur: cursor(<k1,...,kn-1>, kn)
+	mut: insert(v) | delete | assign(v)   v: VAL
+)
+```
+
+where the `id` is a Lamport timestamp which identifies the operation uniquely; 
+`deps` is a set of casual dependencies of the operation; `cur` describes the 
+position of the document to be modified; and `mut` describes the modification 
+itself.
+
+- **Id**
+
+The operation `id` is a Lamport timestamp (see appendix), which guarantees that
+every operation is uniquely identified globally and that the peers implicitly
+agree on the partial casual order of when the operations occurred, without any 
+other synchronization mechanism.
+
+- **Cursor**
+
+A cursor `cur` describes the path from the root to a branch or leaf of the 
+document in which the mutation will be applied. The cursor must consist only of 
+immutable keys and identifiers.
+
+If we consider the following JSON document
+
+```
+{"shopping": ["cheese", "eggs", "milk"]}
+```
+
+the cursor
+
+```
+cursor(<MapT(doc), listT("shopping")>, id1)
+```
+
+is evaluated to `eggs`
+
+
+**Dependencies**
+
+The dependencies `deps` is a set of identifiers of all operation that the current
+operation depends on. The purpose of the operation dependencies is to ensure 
+that the partial order of the operations is respected. 
+
+In order to maintain the partial ordering of operations, an operation is applied 
+on a local replica only when the local state has applied all the 
+operation dependencies. If an external operation has been received and the set 
+of dependencies is not fulfilled, the received operation is buffered until the 
+dependencies operations are applied to the document.
+
+The dependencies set can be implemented in several ways. The requirements are 
+that `deps` contains information about all the operations that the current 
+operations depends on.
+The operation identifiers which are part of the `deps` set can be represented
+as, for example, a set of Lamport timestamps, version vectors, state vectors or 
+dotted version vectors. 
+
+The implementation decision for which type of identifier to use 
+should take into consideration that the set `deps` may grow impracticable if
+the number of elements in the set grows linearly with the number of
+operations applied to the document.
+
+**Mutation**
+
+A mutation `mut` describes the modification to be applied to the node
+pointed by the `cursor`. The mutation type may be one of `INSERT`, `DELETE`
+and `ASSIGN`. Each document type supports different types of mutation:
+
+|        | MapT | ListT | RegisterT |   |
+|--------|------|-------|-----------|---|
+| INSERT | Yes  | Yes   | No        |   |
+| ASSIGN | No   | Yes   | Yes       |   |
+| DELETE | Yes  | Yes   | Yes       |   |
+
+*Table: Document type and support to type of mutation*
+
+### Generating Operations
+
+### Applying Operations
+
+## Document editing API
```

Comment:
Thinking of having a high level pseudo-code representation of the API here. Is it relevant?

Reply:
I think it is relevant, it will give a concrete sense of the semantics provided to the user.